include/rosidl_dynamic_typesupport/api/dynamic_data.h

```diff
@@ -0,0 +1,600 @@
+#include <rosidl_dynamic_typesupport/uchar.h>
+
+
+// ===============================================================================================
+// DYNAMIC DATA
+// ===============================================================================================
+
+// DYNAMIC DATA UTILS ==============================================================================
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_clear_all_values(
+  rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data);
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_clear_nonkey_values(
+  rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data);
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_clear_value(
+  rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data,
+  rosidl_dynamic_typesupport_member_id_t id);
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_equals(
+  const rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data,
+  const rosidl_dynamic_typesupport_dynamic_data_t * other,
+  bool * equals);  // OUT
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_get_item_count(
+  const rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data,
+  size_t * item_count);  // OUT
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_get_member_id_by_name(
+  const rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data,
+  const char * name, size_t name_length,
+  rosidl_dynamic_typesupport_member_id_t * member_id);  // OUT
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_get_member_id_at_index(
+  const rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data,
+  size_t index,
+  rosidl_dynamic_typesupport_member_id_t * member_id);  // OUT
+
+// You must use this for arrays
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_get_array_index(
+  const rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data,
+  size_t index,
+  rosidl_dynamic_typesupport_member_id_t * array_index);  // OUT
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_loan_value(
+  rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data,
+  rosidl_dynamic_typesupport_member_id_t id,
+  rosidl_dynamic_typesupport_dynamic_data_t ** loaned_dynamic_data);  // OUT
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_return_loaned_value(
+  rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data,
+  rosidl_dynamic_typesupport_dynamic_data_t * inner_dynamic_data);
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rcutils_ret_t
+rosidl_dynamic_typesupport_dynamic_data_print(
+  const rosidl_dynamic_typesupport_dynamic_data_t * dynamic_data);
```

Comment:
We've discussed this offline and agreed to remove these print functions because:

- it isn't necessary to delegate this to the rmw implementation, this can be done by the user
- we shouldn't print directly from a library, instead it should be to_string() if anything at all
- we can add utilities later that don't rely on the rmw implmentation, but we cannot easily remove this function later